proposals/0317-sdl-protocol-security-specification.md

```diff
@@ -370,7 +264,115 @@ The following query header is used by the system and the application to send err
   </tr>
 </table>
 
-#### 4.7.5.4 Actions in case of an error
+
+## 7. Secured Communication
+
+It is possible to establish a secured and encrypted communication with the system by setting the frame header encryption flag to `1` when starting a new service. If the authentication was successful, the system will reply with a `StartService ACK` frame with the encryption flag also set to `1` indicating that encrypted data is now accepted. If the authentication fails for some reason the system will reset the TLS connection and return a `StartService NAK` frame.
+
+The RPC service needs to be started as unencrypted first, then moved to encrypted by sending another `StartService` at a later point. Other services can do the same thing to move from unencrypted to encrypted.
+
+Before the encryption of RPC service is enabled (encryption is not available), SDL Core rejects any RPC request with result code `ENCRYPTION_NEEDED` if the RPC needs protection (please see policy updates for which RPC needs protection). SDL Core continues processing an RPC request if the RPC does not need protection. SDL Core sends a notification only if the notification RPC does not need protection.
+
+After the encryption of RPC service is enabled (encryption is available), SDL Core rejects any unencrypted RPC request with result code `ENCRYPTION_NEEDED` with the unencrypted response if the RPC needs protection. SDL Core continues processing an unencrypted RPC request if the RPC does not need protection and responds with an unencrypted response. SDL Core continues processing an encrypted RPC request if the RPC needs protection and responds with an encrypted response. SDL Core sends an unencrypted notification if the RPC does not need protection. SDL Core sends an encrypted notification if the RPC needs protection. In addition, SDL Core shall continue processing an encrypted RPC request if the RPC does not need protection and responds with an encrypted response.
+
+### 7.1 Authentication
+
+The below diagram shows the sequence of how the TLS handshake exchanges certificates to compute the master secret.
+
+![TLS Handshake activity diagram](../assets/proposals/0317-sdl-protocol-security-specification/tls-handshake.png)
+
+The authentication is done using TLS handshake. The TLS handshake process is defined by TLS and is not part of the SDL protocol. The handshake is designed as a client server communication which is configurable in the system settings. An application should take the role of a server where the system is the client. The client entity will initiate a TLS handshake with the corresponding security manager of the server. The client will do this only if the server was not authenticated before in the current transport connection. According to the TLS handshake process the peer certificate can be omitted for the server but it's required for the client.
```

Comment:
I'm not sure the following is correct:
> According to the TLS handshake process the peer certificate can be omitted for the server but it's required for the client.

As far as I know currently client (SDL Core) verifies server (Mobile) peer certificate only in case if .INI option `VerifyPeer` is set to `true`

From other hand current implementation of mobile libraries (this needs to be checked) doesn't require verification of client (SDL Core) certificate by server (Mobile)

Reply:
@dboltovskyi added part related to configurable parameter in SDL ini - https://github.com/LuxoftSDL/sdl_evolution/pull/7/commits/e463f69d7b47d7280aa2472ab9ab84bc27655b44

TODO - investigate this part from the mobile library side

Reply:
@dboltovskyi got a confirmation from Markos that Core certificate is not required. Added this in https://github.com/LuxoftSDL/sdl_evolution/pull/7/commits/93993acd005255a1f377ba8f7b693d7162536eec